### PR TITLE
Fix Application.Unload crashing on certain devices

### PR DESCRIPTION
**Pull request by NinjaKyat**
Fix an issue where jni::Class instances were deleted twice due to how Unity MemoryManager works. Allow deleting these classes when needed.

Added an expanding array that will be used to track all of the objects that need to be cleaned up at some point.

Added a static cleanup method to properly dispose of class objects when needed (call before MemoryManager is shut down).

---

The crash happens because the current Class objects are scoped to the library - they will always be deleted when unloading the unity library. The class objects also have pointers to objects that are allocated on the heap. This heap is cleared/trashed when MemoryManager is unloaded. The class objects try to delete those heap objects when their memory is already wiped. This is where we crash.

Now, we have a new GlobalClass class, which allocates the Class objects on the heap. It is also responsible for disposing of those objects. We can now destroy all the Class objects before we shut down the memory manager. When unloading the unity library, GlobalClass will no longer try to destroy anything, since it will know that it's Class has already been disposed. Therefore, we will no longer crash.

### CODE REVIEW

File: Proxy.cpp [hunk 1]
```diff
@@ -65,7 +65,11 @@ jobject ProxyObject::__Invoke(jclass clazz, jmethodID mid, jobjectArray args)
 
 void ProxyObject::DeleteAllProxies()
 {
-	proxyTracker.DeleteAllProxies();
+	while (g_AllProxies.GetCount() > 0)
```
Comment (juliusmik) [inline Proxy.cpp]: This looks like an infinite loop

File: ExpandingArray.h [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+namespace jni {
+	template <class T>
+	class ExpandingArray {
+	public:
+		ExpandingArray()
+		{
+			pthread_mutexattr_t attr;
+			pthread_mutexattr_init(&attr);
+			pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE);
```
Comment (aurimasc) [inline ExpandingArray.h]: Why mutex is recursive? I don't see guarded methods calling each other.

File: ExpandingArray.h [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+		int count = 0;
+		int maxSize = 32;
+		pthread_mutex_t lock;
+
+		void ExpandArray()
+		{
+			auto biggerArray = new T[maxSize*2];
+			for (int i = 0; i < maxSize; i++)
+			{
+				biggerArray[i] = elements[i];
```
Comment (aurimasc) [inline ExpandingArray.h]: memcpy would probably be better.

File: ExpandingArray.h [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+			pthread_mutex_lock(&lock);
+			T targetElement = elements[index];
+			pthread_mutex_unlock(&lock);
+			return targetElement;
+		}
+
+	private:
+		T* elements;
+		int count = 0;
+		int maxSize = 32;
```
Comment (aurimasc) [inline ExpandingArray.h]: Better rename this to "capacity". maxSize has a feel of hard limit, while capacity is the name used by STL containers.

File: ExpandingArray.h [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+		void Remove(T target)
+		{
+			pthread_mutex_lock(&lock);
+			for (int i = count - 1; i >= 0; i--)
+			{
+				if (elements[i] == target)
+				{
+					if (i != count - 1)
+						elements[i] = elements[count-1];
```
Comment (aurimasc) [inline ExpandingArray.h]: Please add a comment that Remove() swaps the removee with last element prior to removal. Too a while for me to notice.

File: APIHelper.cpp [hunk 6]
```diff
@@ -6,18 +6,41 @@
...
+	GlobalClass::~GlobalClass()
+	{
+		g_AllClasses.Remove(this);
+		if (m_Class)
+			delete m_Class;
+	}
+
+	void GlobalClass::CleanupAllClasses()
+	{
+		for (int i = 0; i < g_AllClasses.GetCount(); i++)
```
Comment (aurimasc) [inline APIHelper.cpp]: Wouldn't it be more efficient to iterate from the last?

**Comment by aurimasc** [comment]
Have you measured how much of an overhead does this add?

**Comment by NinjaKyat** [comment]
I'll measure and compare the startup/shutdown times before and after my changes.

**Review by bitter — CHANGES_REQUESTED** [review]
An expanding array is the wrong data data structure. A (double) linked list would perform much better.

**Comment by bitter** [comment]
I also have concerns regarding size bloat. By templating ExpandingArray you force the compiler to inline code in all compilation units using Proxy.h or ExpandingArray.h.

**Comment by aurimasc** [comment]
The performance should be profiled, rather than assumed. With expandable array you pay for memory allocation when growing, with linked list - every time (unless some smarter memory management). Plus array is more friendly to CPU cache.

**Comment by NinjaKyat** [comment]
Since the current implementation handles pointers, I doubt I'm winning much with an array over a linked list.

File: LinkedList.h [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+		~LinkedList()
+		{
+			pthread_mutex_destroy(&lock);
+		}
+
+		void Add(T* obj)
+		{
+			pthread_mutex_lock(&lock);
+			head = new Link(obj, head);
```
Comment (bitter) [inline LinkedList.h]: Now that you know you only deal with Proxy and Class you can remove this heap allocation and make Class and Proxy linkable. You do this by moving the next ptr into Class and Proxy. That way you can drop Link completely.

**Review by bitter — APPROVED** [review]

**Review by aurimasc — APPROVED** [review]

File: ExpandingArray.h (hunk 2 above)
Comment (NinjaKyat) [reply ExpandingArray.h]: CleanupAll calls into Cleanup which tries to remove the object from the list.

File: ExpandingArray.h (hunk 3 above)
Comment (NinjaKyat) [reply ExpandingArray.h]: This code no longer exists

File: ExpandingArray.h (hunk 4 above)
Comment (NinjaKyat) [reply ExpandingArray.h]: Ditto, code is no longer valid.

**Comment by NinjaKyat** [comment]
@aurimasc you seem to have reviewed some of the older code, most of which got thrown out. Is the new code okay?

**Pull request merged**